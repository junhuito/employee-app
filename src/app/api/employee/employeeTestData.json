[
  {
    "id": 1,
    "name": "John",
    "email": "john@example.com",
    "status": "active"
  },
  {
    "id": 2,
    "name": "Ali",
    "email": "ali@example.com",
    "status": "active"
  },
  {
    "id": 3,
    "name": "Kelly",
    "email": "kelly@example.com",
    "status": "deactivated"
  }
]
