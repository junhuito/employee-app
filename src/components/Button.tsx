import { ButtonHTMLAttributes, DetailedHTMLProps } from "react";

type ButtonDefaultHTMLAttributes = DetailedHTMLProps<
  ButtonHTMLAttributes<HTMLButtonElement>,
  HTMLButtonElement
>;

type ButtonTheme = "primary" | "secondary" | "danger";

type Button = {
  text: string;
  theme?: ButtonTheme;
  type?: "button" | "reset" | "submit";
};

export const Button = (props: ButtonDefaultHTMLAttributes & Button) => {
  const {
    text,
    theme = "primary",
    type = "button",
    ...buttonDefaultHTMLAttributes
  } = props;

  const buttonTheme: Record<ButtonTheme, string> = {
    primary: 'bg-btn-primary-700 hover:bg-btn-primary-500',
    secondary: 'bg-btn-secondary-700 hover:bg-btn-secondary-500',
    danger: 'bg-btn-danger-700 hover:bg-btn-danger-500',
  };

  return (
    <button
      {...buttonDefaultHTMLAttributes}
      type={type}
      className={`text-white ${buttonTheme[theme]} disabled:cursor-not-allowed cursor-pointer font-medium rounded-lg text-sm px-5 py-2.5 focus:outline-none`}
    >
      {text}
    </button>
  );
};
