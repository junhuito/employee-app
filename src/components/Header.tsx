export const Header = () => {
  return (
    <nav className='border-gray-200 bg-primary-800 text-white'>
      <div className='mx-auto flex max-w-screen-xl flex-wrap items-center justify-between p-4'>
        <a
          href={'./'}
          className='flex items-center space-x-3 rtl:space-x-reverse'
        >
          <span className='bg-gradient-to-r from-sky-400 to-emerald-600 bg-clip-text text-2xl font-extrabold tracking-widest text-transparent'>
            AYP
          </span>
        </a>
      </div>
    </nav>
  );
};
