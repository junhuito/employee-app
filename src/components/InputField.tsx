import { DetailedHTMLProps, HTMLInputTypeAttribute, InputHTMLAttributes } from "react";

type InputText = DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>

type InputField = InputText & {
  label: string;
  type?: HTMLInputTypeAttribute;
}

export const InputField = (props: InputField) => {
  const { label, type = "text", ...inputProps } = props;

  return (
    <div>
      <label
        htmlFor="success"
        className="block mb-2 text-sm font-medium text-black dark:text-white"
      >
        {label}
      </label>
      <input
        {...inputProps}
        type={type}
        className="outline-none border-2 border-gray-200 text-sm rounded-lg block w-full p-2.5 text-black"
      />
    </div>
  );
};
