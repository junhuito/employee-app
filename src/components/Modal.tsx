import classnames from 'classnames';

type Modal = {
  children: JSX.Element;
  visible?: boolean;
  onDismiss?: () => void;
};

export const Modal = ({
  onDismiss = () => void 0,
  visible = false,
  children,
}: Modal) => {
  return (
    <div
      role='modal'
      onClick={onDismiss}
      className={`${classnames({
        hidden: !visible,
      })} fixed !inset-0 z-50 !m-0 flex items-center justify-center bg-black bg-opacity-60`}
    >
      <div onClick={(e) => e.stopPropagation()}>{children}</div>
    </div>
  );
};
