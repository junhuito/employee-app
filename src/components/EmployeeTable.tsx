import { Button } from "./Button";
import { EmployeeData } from "@/interfaces/employee.interface";

export type EmployeeTable = {
  data: EmployeeData[];
  onActionClick?: (callback: EmployeeData) => void;
};

export const EmployeeTable = ({
  data = [],
  onActionClick = () => void 0,
}: EmployeeTable) => {
  return (
    <div className="relative overflow-x-auto shadow-md sm:rounded-lg">
      <table className="w-full text-sm md:table-fixed text-left rtl:text-right text-gray-500 dark:text-gray-400">
        <thead className="text-xs text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-400">
          <tr>
            <th scope="col" className="px-6 py-3">
              ID
            </th>
            <th scope="col" className="px-6 py-3">
              Name
            </th>
            <th scope="col" className="px-6 py-3">
              Email
            </th>
            <th scope="col" className="px-6 py-3">
              Status
            </th>
            <th scope="col" className="px-6 py-3">
              <span className="sr-only">Action</span>
            </th>
          </tr>
        </thead>
        <tbody>
          {data.map((employeeData) => (
            <tr
              key={employeeData.id}
              className="bg-white border-b dark:bg-gray-800 dark:border-gray-700 hover:bg-gray-50 dark:hover:bg-gray-600"
            >
              <th
                scope="row"
                className="px-6 py-4 font-medium text-gray-900 whitespace-nowrap dark:text-white"
              >
                {employeeData.id}
              </th>
              <td className="px-6 py-4">{employeeData.name}</td>
              <td className="px-6 py-4">{employeeData.email}</td>
              <td className="px-6 py-4 capitalize">{employeeData.status}</td>
              <td className="px-6 py-4 text-right">
                <Button
                  text="Update"
                  onClick={(e) => onActionClick(employeeData)}
                />
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};
