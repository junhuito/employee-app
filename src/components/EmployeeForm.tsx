import { ChangeEvent, FormEvent, useState } from 'react';
import { Button } from './Button';
import { InputField } from './InputField';
import { Toggle } from './Toggle';
import { EmployeeStatus } from '@/constants/employee.enum';
import { EmployeeData } from '@/interfaces/employee.interface';

type EmployeeForm = {
  onSubmit?: (callback: EmployeeData) => void;
  onCancelClick?: () => void;
  data: EmployeeData;
};

export const EmployeeForm = (props: EmployeeForm) => {
  const [employeeData, setEmployeeData] = useState<EmployeeData>(props.data);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setEmployeeData((prev) => ({ ...prev, [name]: value }));
  };

  const handleToggle = (toggleStatus: boolean) => {
    setEmployeeData((prev) => ({
      ...prev,
      status: toggleStatus ? EmployeeStatus.ACTIVE : EmployeeStatus.DEACTIVATED,
    }));
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    props.onSubmit && props.onSubmit(employeeData);
  };

  return (
    <form
      onSubmit={handleSubmit}
      className='space-y-5 rounded-lg bg-white p-10 dark:bg-primary-700'
    >
      <InputField
        label='Name'
        name='name'
        onChange={handleChange}
        placeholder='name'
        required
        value={employeeData.name}
      />
      <InputField
        type='email'
        name='email'
        label='Email'
        onChange={handleChange}
        placeholder='johndoe@sample.com'
        required
        value={employeeData.email}
      />
      <div className='block w-full'>
        <Toggle
          label={employeeData.status}
          labelCapitilize={true}
          checked={employeeData.status === EmployeeStatus.ACTIVE}
          onChange={handleToggle}
        />
      </div>
      <div className='flex justify-between'>
        <Button
          text='Cancel'
          theme='danger'
          onClick={(e) => props.onCancelClick && props.onCancelClick()}
        />
        <Button type='submit' text='Submit' />
      </div>
    </form>
  );
};
