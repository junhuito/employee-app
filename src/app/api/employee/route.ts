import { NextRequest } from 'next/server';
import employeeTestData from './employeeTestData.json';
import { Direction, SortAttributes } from '@/components/EmployeeTable';
import { EmployeeStatus } from '@/constants/employee.enum';

export async function GET(request: NextRequest) {
  const url = new URL(request.url);
  const attribute = url.searchParams.get('sort') as SortAttributes;
  const direction = url.searchParams.get('direction') as Direction;
  const search = url.searchParams.get('search');

  const statusNormalized = (attribute: string) => {
    const statusMapping: Record<string, number> = {
      [EmployeeStatus.ACTIVE]: 0,
      [EmployeeStatus.DEACTIVATED]: 1,
    };

    return statusMapping[attribute] ?? attribute;
  };

  if (search) {
    employeeTestData.employees = employeeTestData.employees.filter(
      (employee) => {
        const str = Object.values(employee).join(' ');
        return str.search(/search/i);
      }
    );
  }

  if (attribute) {
    employeeTestData.employees.sort((a, b) => {
      const anyA = statusNormalized(a[attribute]) as any;
      const anyB = statusNormalized(b[attribute]) as any;

      if (typeof anyA === 'string') {
        return direction === 'ASC'
          ? anyA.localeCompare(anyB)
          : anyB.localeCompare(anyA);
      }

      return direction === 'ASC' ? anyA - anyB : anyB - anyA;
    });
  }

  await new Promise((res) => setTimeout(res, 500));

  return Response.json({ data: employeeTestData.employees });
}
