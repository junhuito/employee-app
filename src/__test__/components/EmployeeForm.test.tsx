import { render } from '@testing-library/react';
import { EmployeeForm } from '@/components/EmployeeForm';
import { EmployeeStatus } from '@/constants/employee.enum';
import '@testing-library/jest-dom';

describe('EmployeeForm', () => {
  it('should render form', () => {
    const { container } = render(
      <EmployeeForm
        data={{
          id: 1,
          name: 'test employee',
          email: 'test@example.com',
          status: EmployeeStatus.ACTIVE,
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
