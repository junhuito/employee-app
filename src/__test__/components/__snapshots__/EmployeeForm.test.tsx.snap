// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeForm should render form 1`] = `
<div>
  <form
    class="space-y-5 rounded-lg bg-white p-10 dark:bg-primary-700"
  >
    <div>
      <label
        class="mb-2 block text-sm font-medium text-black dark:text-white"
      >
        Name
      </label>
      <input
        class="block w-full rounded-lg border-2 border-gray-200 p-2.5 text-sm text-black outline-none"
        name="name"
        placeholder="name"
        required=""
        type="text"
        value="test employee"
      />
    </div>
    <div>
      <label
        class="mb-2 block text-sm font-medium text-black dark:text-white"
      >
        Email
      </label>
      <input
        class="block w-full rounded-lg border-2 border-gray-200 p-2.5 text-sm text-black outline-none"
        name="email"
        placeholder="johndoe@sample.com"
        required=""
        type="email"
        value="test@example.com"
      />
    </div>
    <div
      class="block w-full"
    >
      <label
        class="relative inline-flex cursor-pointer items-center"
      >
        <input
          checked=""
          class="peer sr-only"
          type="checkbox"
        />
        <div
          class="peer h-6 w-11 rounded-full bg-slate-900 after:absolute after:start-[2px] after:top-[2px] after:h-5 after:w-5 after:rounded-full after:border after:border-gray-300 after:bg-white after:transition-all after:content-[''] peer-checked:bg-green-600 peer-checked:after:translate-x-full peer-checked:after:border-white peer-focus:outline-none rtl:peer-checked:after:-translate-x-full"
        />
        <span
          class="ms-3 select-none text-sm font-medium capitalize"
        >
          active
        </span>
      </label>
    </div>
    <div
      class="flex justify-between"
    >
      <button
        class="text-white bg-btn-danger-700 hover:bg-btn-danger-500 cursor-pointer rounded-lg px-2 py-2.5 text-xs font-medium focus:outline-none disabled:cursor-not-allowed"
        type="button"
      >
        Cancel
      </button>
      <button
        class="text-white bg-btn-primary-700 hover:bg-btn-primary-500 cursor-pointer rounded-lg px-2 py-2.5 text-xs font-medium focus:outline-none disabled:cursor-not-allowed"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
